### PR TITLE
Fix and speedup builds with TSan

### PR DESCRIPTION
**Pull request by shindere**
This PR fixes issue #13042 and also addresses the problem PR #12781 attempted to resolve.

Both of these are instances of artefacts for bytecode programs or libraries being instrumented whereas they should not be.

The present PR makes sure only the artefacts related to native programs and libraries get instrumented. As a side effect, the build time for the compiler when configured with `--endable-tsan` is considerably reduced, as can be seen from these results obtained with `make -j`:

Trunk:

```
real    6m27,551s
user    41m52,165s
sys     7m14,887s
```

PR branch:

```
real    3m18,914s
user    18m52,346s
sys     1m57,559s
```

This PR consists of the four following commits:

 1. Variable renamings local to the `configure` script.
 2. Make sure `runtime/tsan.c` is compiled and linked only for native programs and libraries.
 3. Restoring the possibility to use different C and C preprocessor flags when compiling C programs in bytecode and native contexts.
 4. Taking advantage of the previous commit to make sure only the bytecode-related artefacts get instrumented.

Of these four commits, the third one can certainly be seen as the core of this PR. It basically reverts 2b4fe09dc7196a4814710495e29a183dbeabad06, which had been submitted as part of PR #8631. However, the present PR proposes new names for the different types of flags, which are based on the backend for which we compile rather than on the name of the associated compiler, since the backend names seem more stable than the compiler names. So we use `bytecode_cflags` rather than `ocamlc_cflags`, etc., for the day when `ocamlc` and `ocamlopt` will have a unified compiler driver. ;)

This work may be of interest to @fabbing and @OlivierNicole, among others.

### CODE REVIEW

**Review by ghost — APPROVED** [review]
I was worried about the removal of `-fsanitize=thread` from `cclibs` in `configure.ac` (last chunk), but it turns out to be indeed unnecessary as long as `*cflags` are correct.

**Comment by gasche** [comment]
There are weird mscv64 failure logs in the CI, and I don't know if they are a fluke (a bug in the mscv64 CI in trunk at the time, that was fixed since) or related to the PR. @shindere, could you rebase the PR on top of the current trunk, so that the CI reruns in a known-good configuration?

**Comment by shindere** [comment]
I will rebase yes, and intend to go through the PR again because I think
there is still room for improvemnet. In particular, I think it's
important to pass the TSan flags at link time because, as I understood
it, some platforms do require that.

**Comment by ghost** [comment]
> I will rebase yes, and intend to go through the PR again because I think there is still room for improvemnet. In particular, I think it's important to pass the TSan flags at link time because, as I understood it, some platforms do require that.

All platforms do - TSan-enabled binaries need to be linked against the TSan runtime library.

**Comment by shindere** [comment]
I just pushed an updated version of the PR that passes CI.

(I know it does because it did and I just squashed a tentative fix in the
commit it was actually fixing.)

Basically, the problem was that there were some flags missing in those
ocamlopt must pass to the C files it compiles to be linked with native code.

This is, in turn, the consequence of two things. On the one hand, it is
currently true that the C flags used for native code are a superset of those
used for bytecode. On the other hand, this assumption, which is true
currently but may not always be true, is expressed at several places in the
build system, namely everytime we say that the native flags are the bytecode
flags plus something else.

I think this should be fixed and the build system shoudl not compute the
native code flags in terms of the bytecode ones. It should rather use the
native code flags, whatever they are, and those should be computed
once and for all during the configure stage, where we can compute
them in connection with the bytecode flags as appropriate.
But that's quite a bit of work and quite orthogonal to the current
PR so I would like to leave it for a dedicated PR which I intend to
submit soon.

File: configure.ac [hunk 1]
```diff
@@ -2339,6 +2340,7 @@ AS_CASE([$host],
     [PTHREAD_LIBS=''],
   [AX_PTHREAD(
     [common_cflags="$common_cflags $PTHREAD_CFLAGS"
+    AC_MSG_NOTICE([SEB: common_cflags="$common_cflags"])
```
Comment (ghost) [inline configure.ac]: You forgot that leftover debug message.
```suggestion
```

**Comment by shindere** [comment]
Miod Vallat (2024/06/03 02:52 -0700):
> @dustanddreams commented on this pull request.
>
>
>
> > @@ -2339,6 +2340,7 @@ AS_CASE([$host],
>      [PTHREAD_LIBS=''],
>    [AX_PTHREAD(
>      [common_cflags="$common_cflags $PTHREAD_CFLAGS"
> +    AC_MSG_NOTICE([SEB: common_cflags="$common_cflags"])
>
> You forgot that leftover debug message.
> ```suggestion
> ```

Oops! Thanks! Fixed!

File: Makefile.config.in [hunk 2]
```diff
@@ -175,10 +175,21 @@ NATDYNLINK=@natdynlink@
...
+# link with bytecode.
+
+# These flags should be passed *in addition* to those in OC_CPPFLAGS, they
+# should not replace them.
+
+OC_NATIVE_CPPFLAGS=@native_cppflags@
+
+# Same as above, for CFLAGS
+OC_NATIVE_CFLAGS=@native_cflags@
+
```
Comment (OlivierNicole) [inline Makefile.config.in:192]: As I understand it from private discussion, these two sets of flags cannot currently be complemented by the user, but may be in the future (which is fine).

**Review by OlivierNicole — APPROVED** [review]
This looks good to me too (after some discussion with the author to learn more about the workings of autoconf and some naming conventions).

I’m also very happy to speed up TSan and above all get rid of those CI false positives.

**Comment by shindere** [comment]
Thanks a lot @oliviernicole for your very thorough review.

Olivier has asked very good quesitons and I was happy to have the
possibility to explain in a ocnversation rather than in a written form,
until we hae a good documentaiton for all the build system.

**Comment by shindere** [comment]
Gabriel Scherer (2024/06/04 04:51 -0700):
> Merged #13201 into trunk.

Thnaks!

**Comment by anmonteiro** [comment]
> So we use bytecode_cflags rather than ocamlc_cflags, etc., for the day when ocamlc and ocamlopt will have a unified compiler driver. ;)

Could someone provide a stronger argument for breaking every single user of `ocamlc -config` starting in 5.3 other than the hand-wavy "look, guys, it'll be better when the compiler driver is unified"?

**Comment by gasche** [comment]
I think that you are reporting an unintended regression, but there is too much sarcasm and not enough information to know what the regression is and how to fix it. Could you try again with a gentler explanation of the issue?

**Comment by anmonteiro** [comment]
No sarcasm intended, this is a pretty serious regression. On OCaml 5.2:

```
$ ocamlc.opt -config
...
ocamlc_cflags:  -O2 -fno-strict-aliasing -fwrapv -pthread   -pthread
ocamlc_cppflags:  -D_FILE_OFFSET_BITS=64
ocamlopt_cflags:  -O2 -fno-strict-aliasing -fwrapv -pthread   -pthread
ocamlopt_cppflags:  -D_FILE_OFFSET_BITS=64
...
```

On OCaml trunk (what will be 5.3):

```
$ ocamlc.opt -config
...
bytecode_cflags:  -O2 -fno-strict-aliasing -fwrapv -pthread   -pthread
bytecode_cppflags:  -D_FILE_OFFSET_BITS=64
native_cflags:  -O2 -fno-strict-aliasing -fwrapv -pthread   -pthread
native_cppflags:  -D_FILE_OFFSET_BITS=64  -DNATIVE_CODE -DTARGET_arm64 -DMODEL_default -DSYS_macosx
...
```

Tooling such as dune [relies](https://github.com/ocaml/dune/blob/987e1f16b3e6a9d43ec7e3cb130fe52ec86fceca/src/ocaml-config/ocaml_config.ml#L558-L561) on those variables being present.

**Comment by gasche** [comment]
Thanks! This sounds easy to fix by undoing the renaming from `ocamlc, ocamlopt` to `bytecode, native`, which is pleasant but not worth any userland breakage.

**Comment by gasche** [comment]
(I'm not planning to work on this myself right away as I suppose that @shindere or @OlivierNicole will probably do it, but of course anyone else is welcome to propose a PR.)

**Comment by anmonteiro** [comment]
If there's a strong desire to rename theses variables, you can still introduce the new variables alongside the current ones and establish some kind of deprecation strategy. 

That's what puzzled me so much, given the OCaml compiler's track record of not introducing breaking changes like these.

**Comment by shindere** [comment]
Antonio Nuno Monteiro (2024/06/14 21:43 -0700):
> If there's a strong desire to rename theses variables, you can still
> introduce the new variables alongside the current ones and establish
> some kind of deprecation strategy.

Taht was actually my initial intention, and then I forgot about it, I
apologize for that and the trouble it created.

> That's what puzzled me so much, given the OCaml compiler's track
> record of not introducing breaking changes like these.

So is it okay if I (1) submit a PR that re-introduces the older names
but with deprecation marks and (2) make sure dune is transitionning to
the new names?

I intend to submit this as part of a revised version of #12586 I am
currently working on, unless people feel there is a need for a dedicated
PR but I'd definitely prefer to have this in #12586 which basically
deals with C and CPP flags.

**Comment by gasche** [comment]
I think that we should just go back to the old names. `ocamlc` and `ocamlopt` are clear enough and the new names are marginally nicer, but not worth the work of deprecating and communicating about it and asking our users to change their practices and and... This is a standard "perfect is the enemy of good" situation.

> unless people feel there is a need for a dedicated PR

The renaming would be a very small change that can easily be reviewed and merged, and un-break `trunk` for our downstream users. I would be in favor of doing this now instead of tying it to something bigger.

**Comment by shindere** [comment]
I'd like to insist on the renaming because it's more future proof.

As I think I already wrote, I am hoping that we will once have a unified
compiler that will have several backends, including bytecode and native.
To me the proposed names work better in such a context and are easier to
extend to potential other backends.

**Comment by gasche** [comment]
In my opinion the priority is to un-break the various CI systems that are following trunk to make the ecosystem better ahead of time, with the minimal change that achieves it. I can either (1) just let you do whatever you want, even if I don't like the result, because other approaches lead to potential conflicts, or (2) do the minimal thing myself. I guess that a last option is to (3) revert the present PR from trunk for the time being. What do you think is best?

**Comment by shindere** [comment]
Am I correct that re-introducing the old names with deprecation would fix
trunk? If so I can prepare a PR with just that promptly.

**Comment by shindere** [comment]
Just opened #13244 to fix `ocamlc -config`

**Pull request merged**